File: app/src/main/java/com/erencol/sermon/view/activities/MainActivity.java
package com.erencol.sermon.view.activities;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.BillingClientStateListener;
import com.android.billingclient.api.BillingResult;
import com.android.billingclient.api.PendingPurchasesParams;
import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.PurchasesUpdatedListener;
import com.android.billingclient.api.QueryPurchasesParams;
import com.erencol.sermon.Data.service.manager.PremiumManager;
import com.erencol.sermon.view.adapters.SermonAdapter;
import com.erencol.sermon.databinding.ActivityMainBinding;
import com.erencol.sermon.viewmodelpkg.MainViewModel;
import java.util.List;
import java.util.Observable;
import java.util.Observer;
import com.erencol.sermon.R;

public class MainActivity extends AppCompatActivity implements Observer, PurchasesUpdatedListener {
    private MainViewModel mainViewModel;
    private ActivityMainBinding binding;
    private BillingClient billingClient;
    private PremiumManager premiumManager;
    private SermonAdapter sermonAdapter;
    private static final String PREMIUM_SKU = "premium"; // Google Play Console'da tanımladığınız SKU ID'si
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initDataBinding();
        initBillingClient();
        checkisPremium();
    }

    private void initBillingClient() {
        PendingPurchasesParams pendingPurchasesParams = PendingPurchasesParams.newBuilder()
                .enableOneTimeProducts()
                .build();
        
        billingClient = BillingClient.newBuilder(this)
                .setListener(this)
                .enablePendingPurchases(pendingPurchasesParams)
                .build();
        
        billingClient.startConnection(new BillingClientStateListener() {
            @Override
            public void onBillingSetupFinished(BillingResult billingResult) {
                if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
                    queryPurchases();
                }
            }

            @Override
            public void onBillingServiceDisconnected() {
                // Billing servisi bağlantısı kesildi, tekrar bağlanmayı dene
            }
        });
    }

    private void queryPurchases() {
        billingClient.queryPurchasesAsync(
            QueryPurchasesParams.newBuilder()
                .setProductType(BillingClient.ProductType.INAPP)
                .build(),
            (billingResult, purchases) -> {
                if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
                    handlePurchases(purchases);
                }
            }
        );
    }

    private void handlePurchases(List<Purchase> purchases) {
        boolean isPremium = false;
        for (Purchase purchase : purchases) {
            if (purchase.getProducts().contains(PREMIUM_SKU) && purchase.getPurchaseState() == Purchase.PurchaseState.PURCHASED) {
                isPremium = true;
                break;
            }
        }
        premiumManager.setPremium(isPremium);
        updateAdapterPremiumStatus();
    }

    public void checkisPremium(){
        // Premium durumu adapter'a geçirilecek
        updateAdapterPremiumStatus();
    }
    
    private void updateAdapterPremiumStatus() {
        if (sermonAdapter != null) {
            sermonAdapter.setPremium(premiumManager.isPremium());
        }
    }

    public void initDataBinding(){
        premiumManager = PremiumManager.getInstance(getApplicationContext());
        binding = DataBindingUtil.setContentView(this,R.layout.activity_main);
        mainViewModel = new MainViewModel();
        binding.setMainViewModel(mainViewModel);
        binding.setLifecycleOwner(this);
        setSupportActionBar(binding.toolbar);
        setListSermonListview(binding.sermonsRecyclerview);
        setupObserver(mainViewModel);
    }

    private void showPremiumDialog() {
        new AlertDialog.Builder(this)
                .setTitle("Premium İçerik")
                .setMessage("Bu içerik sadece premium kullanıcılar için. Premium sürümü satın almak ister misiniz?")
                .setPositiveButton("Satın Al", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {

                    }
                })
                .setNegativeButton("İptal", null)
                .show();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main,menu);
        return true;
    }

    public void setupObserver(Observable observable) {
        observable.addObserver(this);
    }
    private void setListSermonListview(RecyclerView listPeople) {
        sermonAdapter = new SermonAdapter();
        sermonAdapter.setPremium(premiumManager.isPremium());
        listPeople.setAdapter(sermonAdapter);
        listPeople.setLayoutManager(new LinearLayoutManager(this));
        mainViewModel.getSermons();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.aboutid) {
            showPremiumDialog();
            //showAboutActivity();
            return true;
        } else if (id == R.id.specialdaysid) {
            goToSpecialDaysActivity();
            return true;
        } else if (id == R.id.settingsid) {
            goToSettingsActivity();
            return true;
        } /* else if (id == R.id.qiblaId) {
            goToQibla();
            return true;
        }*/
        else {
            return super.onOptionsItemSelected(item);
        }
    }

    public void goToSpecialDaysActivity(){
        Intent i = new Intent(MainActivity.this, SpecialDaysActivity.class);
        startActivity(i);
    }

    public void goToSettingsActivity(){
        Intent i = new Intent(MainActivity.this,SettingsActivity.class);
        startActivity(i);
    }

    public void showAboutActivity(){
        Intent i = new Intent(MainActivity.this, AboutActivity.class);
        startActivity(i);
    }

    public void goToQibla(){
        Intent i = new Intent(MainActivity.this, CompassActivity.class);
        startActivity(i);
    }

    @Override public void update(Observable observable, Object data) {
        if (observable instanceof MainViewModel mainViewModel) {
            SermonAdapter adapter = (SermonAdapter) binding.sermonsRecyclerview.getAdapter();
            assert adapter != null;
            adapter.setSermonList(mainViewModel.getSermonList().getValue());
            adapter.setPremium(premiumManager.isPremium());
        }
    }
    
    @Override
    public void onPurchasesUpdated(BillingResult billingResult, List<Purchase> purchases) {
        if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK && purchases != null) {
            handlePurchases(purchases);
        }
    }
    
    @Override
    protected void onResume() {
        super.onResume();
        // Her activity'ye dönüldüğünde premium durumunu kontrol et
        if (billingClient != null && billingClient.isReady()) {
            queryPurchases();
        }
    }
    
    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (billingClient != null) {
            billingClient.endConnection();
        }
    }


}

File: app/src/main/java/com/erencol/sermon/view/main/MainActivity.kt
package com.erencol.sermon.view.main

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.net.Uri
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import com.erencol.sermon.R
import com.erencol.sermon.data.manager.SharedPreferencesManager
import com.erencol.sermon.databinding.ActivityMainBinding
import com.erencol.sermon.fcm.FirebaseTopicManager
import com.erencol.sermon.view.about.AboutActivity
import com.erencol.sermon.view.settings.SettingsActivity
import com.erencol.sermon.view.specialdays.SpecialDaysActivity
import java.util.Observable
import java.util.Observer

class MainActivity : AppCompatActivity(), Observer {
    private lateinit var mainViewModel: MainViewModel
    private lateinit var binding: ActivityMainBinding
    private lateinit var billingManager: com.erencol.sermon.billing.BillingManager
    private lateinit var firebaseTopicManager: FirebaseTopicManager

    companion object {
        private const val TAG = "MainActivity"
        private const val NOTIFICATION_PERMISSION_REQUEST_CODE = 1001
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initDataBinding()
        checkRateUsDialog()
    }

    private fun initDataBinding() {
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        mainViewModel = MainViewModel()
        binding.mainViewModel = mainViewModel
        binding.lifecycleOwner = this
        setSupportActionBar(binding.toolbar)
        
        setupBilling()
        setupFirebaseMessaging()
        setListSermonListview()
        setupObserver(mainViewModel)
        setupFab()
    }

    private fun setupFab() {
        binding.fabPremium.setOnClickListener {
            androidx.appcompat.app.AlertDialog.Builder(this)
                .setTitle(getString(R.string.premium_popup_title))
                .setMessage(getString(R.string.premium_popup_message))
                .setPositiveButton(getString(R.string.analyze)) { _, _ ->
                    billingManager.launchPurchaseFlow(this)
                }
                .setNegativeButton(getString(R.string.later), null)
                .show()
        }
    }

    private fun setupBilling() {
        // FirebaseTopicManager'ı önce oluştur
        firebaseTopicManager = FirebaseTopicManager(this)
        
        // BillingManager'a FirebaseTopicManager'ı geç
        billingManager = com.erencol.sermon.billing.BillingManager(this, firebaseTopicManager)
        billingManager.startConnection()
        billingManager.isPremium.observe(this, { isPremium ->
            val adapter = binding.sermonsRecyclerview.adapter as? SermonAdapter
            adapter?.isPremium = isPremium
            binding.fabPremium.visibility = if (isPremium) View.GONE else View.VISIBLE
        })
    }

    private fun setupFirebaseMessaging() {
        // Android 13+ için bildirim izni kontrolü
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (ContextCompat.checkSelfPermission(
                    this,
                    Manifest.permission.POST_NOTIFICATIONS
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                ActivityCompat.requestPermissions(
                    this,
                    arrayOf(Manifest.permission.POST_NOTIFICATIONS),
                    NOTIFICATION_PERMISSION_REQUEST_CODE
                )
            } else {
                Log.d(TAG, "Notification permission already granted")
            }
        } else {
            // Android 13'ten önceki sürümler için izin gerekmiyor
            Log.d(TAG, "Notification permission not required for this Android version")
        }
        
        // FCM Token'ı logla (debug için)
        firebaseTopicManager.getFcmToken { token ->
            Log.d(TAG, "FCM Token: $token")
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        
        if (requestCode == NOTIFICATION_PERMISSION_REQUEST_CODE) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Log.d(TAG, "Notification permission granted")
            } else {
                Log.d(TAG, "Notification permission denied")
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    private fun setupObserver(observable: Observable) {
        observable.addObserver(this)
    }

    private fun setListSermonListview() {
        val adapter = SermonAdapter()
        adapter.onPremiumContentClick = {
             androidx.appcompat.app.AlertDialog.Builder(this)
                .setTitle(getString(R.string.premium_popup_title))
                .setMessage(getString(R.string.premium_popup_message))
                .setPositiveButton(getString(R.string.analyze)) { _, _ ->
                    billingManager.launchPurchaseFlow(this)
                }
                .setNegativeButton(getString(R.string.later), null)
                .show()
        }
        binding.sermonsRecyclerview.adapter = adapter
        binding.sermonsRecyclerview.layoutManager = LinearLayoutManager(this)
        mainViewModel.getSermons()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.aboutid) {
            showAboutActivity()
            return true
        } else if (id == R.id.specialdaysid) {
            goToSpecialDaysActivity()
            return true;
        } else if (id == R.id.settingsid) {
            goToSettingsActivity()
            return true
        } else {
            return super.onOptionsItemSelected(item)
        }
    }

    private fun goToSpecialDaysActivity() {
        val i = Intent(this@MainActivity, SpecialDaysActivity::class.java)
        startActivity(i)
    }

    private fun goToSettingsActivity() {
        val i = Intent(this@MainActivity, SettingsActivity::class.java)
        startActivity(i)
    }

    private fun showAboutActivity() {
        val i = Intent(this@MainActivity, AboutActivity::class.java)
        startActivity(i)
    }

    override fun update(observable: Observable, data: Any?) {
        if (observable is MainViewModel) {
            val sermonAdapter = binding.sermonsRecyclerview.adapter as SermonAdapter?
            sermonAdapter?.setSermonList(observable.sermonList.value)
        }
    }

    private fun checkRateUsDialog() {
        val prefs = SharedPreferencesManager.getInstance(this)
        val currentVersionCode = try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                packageManager.getPackageInfo(packageName, 0).longVersionCode.toInt()
            } else {
                @Suppress("DEPRECATION")
                packageManager.getPackageInfo(packageName, 0).versionCode
            }
        } catch (e: Exception) {
            0
        }

        val lastVersionCode = prefs.getIntForKey(SharedPreferencesManager.KEY_LAST_VERSION_CODE, 0)
        val launchCount = prefs.getIntForKey(SharedPreferencesManager.KEY_LAUNCH_COUNT, 0)
        val shownKey = SharedPreferencesManager.KEY_RATE_SHOWN_PREFIX + currentVersionCode
        val isShownForThisVersion = prefs.getBoolean(shownKey, false)

        if (isShownForThisVersion) return

        if (lastVersionCode == 0) {
            // First time install initialization
            prefs.putIntForKey(SharedPreferencesManager.KEY_LAST_VERSION_CODE, currentVersionCode)
            prefs.putIntForKey(SharedPreferencesManager.KEY_LAUNCH_COUNT, 1)
        } else if (lastVersionCode == currentVersionCode) {
            // Ongoing session for the same version
            val newCount = launchCount + 1
            prefs.putIntForKey(SharedPreferencesManager.KEY_LAUNCH_COUNT, newCount)
            if (newCount == 3) {
                showRateUsDialog(shownKey)
            }
        } else if (lastVersionCode < currentVersionCode) {
            // Updated user
            showRateUsDialog(shownKey)
            prefs.putIntForKey(SharedPreferencesManager.KEY_LAST_VERSION_CODE, currentVersionCode)
        }
    }

    private fun showRateUsDialog(shownKey: String) {
        val view = layoutInflater.inflate(R.layout.dialog_rate_us, null)
        val textView = view.findViewById<android.widget.TextView>(R.id.textViewMessage)
        textView.text = getString(R.string.rate_us_message, getString(R.string.app_name))

        androidx.appcompat.app.AlertDialog.Builder(this)
            .setView(view)
            .setPositiveButton(getString(R.string.ok)) { _, _ ->
                SharedPreferencesManager.getInstance(this).putBoolean(shownKey, true)
                openPlayStore()
            }
            .setNegativeButton(getString(R.string.cancel)) { dialog, _ ->
                SharedPreferencesManager.getInstance(this).putBoolean(shownKey, true)
                dialog.dismiss()
            }
            .setCancelable(false)
            .show()
    }

    private fun openPlayStore() {
        val intent = Intent(Intent.ACTION_VIEW).apply {
            data = Uri.parse("market://details?id=$packageName")
        }
        try {
            startActivity(intent)
        } catch (e: Exception) {
            val webIntent = Intent(Intent.ACTION_VIEW).apply {
                data = Uri.parse("https://play.google.com/store/apps/details?id=$packageName")
            }
            startActivity(webIntent)
        }
    }
}